storage/compactor: TestCompactorDisabled failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/835cc098da0698ea9a12271d4e2998fad69a8e4a

Parameters:
```
TAGS=
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=656006&tab=buildLog

```
	compactor_test.go:690: condition failed to evaluate within 45s: expected skipped bytes 805306365; got 1342177275
		goroutine 86 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc4204fef78, 0x15bebe0)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x15d7d80, 0xc4203700f0, 0xc4204fef78)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:38 +0x7d
		github.com/cockroachdb/cockroach/pkg/storage/compactor.TestCompactorDisabled(0xc4203700f0)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/compactor/compactor_test.go:690 +0x1b7
		testing.tRunner(0xc4203700f0, 0x1523bc8)
			/usr/local/go/src/testing/testing.go:777 +0xd0
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x2e0


ERROR: exit status 1

3555 runs completed, 1 failures, over 1m29s
context canceled
Makefile:869: recipe for target 'stress' failed
make: *** [stress] Error 1
```

--- a/pkg/settings/float.go
+++ b/pkg/settings/float.go
@@ -45,6 +45,14 @@ func (*FloatSetting) Typ() string {
 	return "f"
 }
 
+// Override changes the setting, panicking when validation fails.
+// For testing usage only.
+func (f *FloatSetting) Override(sv *Values, v float64) {
+	if err := f.set(sv, v); err != nil {
+		panic(err)
+	}
+}
+
 // Validate that a value conforms with the validation function.
 func (f *FloatSetting) Validate(v float64) error {
 	if f.validateFn != nil {

--- a/pkg/storage/compactor/compactor.go
+++ b/pkg/storage/compactor/compactor.go
@@ -284,13 +284,24 @@ func (c *Compactor) processCompaction(
 	capacity roachpb.StoreCapacity,
 	delBatch engine.Batch,
 ) (int64, error) {
-	shouldProcess := c.enabled() && (aggr.Bytes >= c.thresholdBytes() ||
-		aggr.Bytes >= int64(float64(capacity.LogicalBytes)*c.thresholdBytesUsedFraction()) ||
-		aggr.Bytes >= int64(float64(capacity.Available)*c.thresholdBytesAvailableFraction()))
+	aboveSizeThresh := aggr.Bytes >= c.thresholdBytes()
+	aboveUsedFracThresh := func() bool {
+		thresh := c.thresholdBytesUsedFraction()
+		return thresh > 0 && aggr.Bytes >= int64(float64(capacity.LogicalBytes)*thresh)
+	}()
+	aboveAvailFracThresh := func() bool {
+		thresh := c.thresholdBytesAvailableFraction()
+		return thresh > 0 && aggr.Bytes >= int64(float64(capacity.Available)*thresh)
+	}()
 
+	shouldProcess := c.enabled() && (aboveSizeThresh || aboveUsedFracThresh || aboveAvailFracThresh)
 	if shouldProcess {
 		startTime := timeutil.Now()
-		log.Infof(ctx, "processing compaction %s", aggr)
+		log.Infof(ctx,
+			"processing compaction %s (reasons: size=%t used=%t avail=%t)",
+			aggr, aboveSizeThresh, aboveUsedFracThresh, aboveAvailFracThresh,
+		)
+
 		if err := c.eng.CompactRange(aggr.StartKey, aggr.EndKey, false /* forceBottommost */); err != nil {
 			c.Metrics.CompactionFailures.Inc(1)
 			return 0, errors.Wrapf(err, "unable to compact range %+v", aggr)
@@ -316,7 +327,7 @@ func (c *Compactor) processCompaction(
 			continue
 		}
 		if tooOld {
-			c.Metrics.BytesSkipped.Inc(aggr.Bytes)
+			c.Metrics.BytesSkipped.Inc(sc.Bytes)
 		}
 		key := keys.StoreSuggestedCompactionKey(sc.StartKey, sc.EndKey)
 		if err := delBatch.Clear(engine.MVCCKey{Key: key}); err != nil {

--- a/pkg/storage/compactor/compactor_test.go
+++ b/pkg/storage/compactor/compactor_test.go
@@ -657,47 +657,60 @@ func TestCompactorCleansUpOldRecords(t *testing.T) {
 func TestCompactorDisabled(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
+	threshold := int64(10)
+
 	capacityFn := func() (roachpb.StoreCapacity, error) {
 		return roachpb.StoreCapacity{
-			LogicalBytes: 100 * thresholdBytes.Default(),
-			Available:    100 * thresholdBytes.Default(),
+			LogicalBytes: 100 * threshold,
+			Available:    100 * threshold,
 		}, nil
 	}
 	compactor, we, compactionCount, cleanup := testSetup(capacityFn)
 	minInterval.Override(&compactor.st.SV, time.Millisecond)
 	maxSuggestedCompactionRecordAge.Override(&compactor.st.SV, 24*time.Hour) // large
+	thresholdBytesAvailableFraction.Override(&compactor.st.SV, 0.0)          // disable
+	thresholdBytesUsedFraction.Override(&compactor.st.SV, 0.0)               // disable
 	defer cleanup()
 
-	suggest := func() {
-		// Add a suggested compaction that won't get processed because it's
-		// not over any of the thresholds.
-		compactor.Suggest(context.Background(), storagebase.SuggestedCompaction{
-			StartKey: key("a"), EndKey: key("b"),
-			Compaction: storagebase.Compaction{
-				Bytes:            thresholdBytes.Default() - 1,
-				SuggestedAtNanos: timeutil.Now().UnixNano(),
-			},
-		})
-	}
+	compactor.Suggest(context.Background(), storagebase.SuggestedCompaction{
+		StartKey: key("a"), EndKey: key("b"),
+		Compaction: storagebase.Compaction{
+			// Suggest so little that this suggestion plus the one below stays below
+			// the threshold. Otherwise this test gets racy and difficult to fix
+			// without remodeling the compactor.
+			Bytes:            threshold / 3,
+			SuggestedAtNanos: timeutil.Now().UnixNano(),
+		},
+	})
 
-	suggest()
-	suggest()
 	enabled.Override(&compactor.st.SV, false)
-	suggest()
+
+	compactor.Suggest(context.Background(), storagebase.SuggestedCompaction{
+		// Note that we don't reuse the same interval above or we hit another race,
+		// in which the compactor discards the first suggestion and wipes out the
+		// second one with it, without incrementing the discarded metric.
+		StartKey: key("b"), EndKey: key("c"),
+		Compaction: storagebase.Compaction{
+			Bytes:            threshold / 3,
+			SuggestedAtNanos: timeutil.Now().UnixNano(),
+		},
+	})
 
 	// Verify that the record is deleted without a compaction and that the
 	// bytes are recorded as having been skipped.
 	testutils.SucceedsSoon(t, func() error {
+		if a, e := atomic.LoadInt32(compactionCount), int32(0); a != e {
+			t.Fatalf("expected compactions processed %d; got %d", e, a)
+		}
+
 		comps := we.GetCompactions()
 		if !reflect.DeepEqual([]roachpb.Span(nil), comps) {
 			return fmt.Errorf("expected nil compactions; got %+v", comps)
 		}
-		if a, e := compactor.Metrics.BytesSkipped.Count(), 3*(thresholdBytes.Get(&compactor.st.SV)-1); a != e {
+		if a, e := compactor.Metrics.BytesSkipped.Count(), 2*(threshold/3); a != e {
 			return fmt.Errorf("expected skipped bytes %d; got %d", e, a)
 		}
-		if a, e := atomic.LoadInt32(compactionCount), int32(0); a != e {
-			return fmt.Errorf("expected compactions processed %d; got %d", e, a)
-		}
+
 		// Verify compaction queue is empty.
 		if bytesQueued, err := compactor.examineQueue(context.Background()); err != nil || bytesQueued > 0 {
 			return fmt.Errorf("compaction queue not empty (%d bytes) or err %v", bytesQueued, err)

--- a/pkg/storage/compactor/settings.go
+++ b/pkg/storage/compactor/settings.go
@@ -78,7 +78,7 @@ var thresholdBytes = settings.RegisterByteSizeSetting(
 // defaultThresholdBytes threshold.
 var thresholdBytesUsedFraction = settings.RegisterValidatedFloatSetting(
 	"compactor.threshold_used_fraction",
-	"Consider suggestions for at least the given percentage of the used logical space",
+	"Consider suggestions for at least the given percentage of the used logical space (zero to disable)",
 	0.10, // more than 10% of space will trigger
 	validateFraction,
 )
@@ -91,7 +91,7 @@ var thresholdBytesUsedFraction = settings.RegisterValidatedFloatSetting(
 // scenarios.	ThresholdBytesAvailableFraction() float64
 var thresholdBytesAvailableFraction = settings.RegisterValidatedFloatSetting(
 	"compactor.threshold_available_fraction",
-	"Consider suggestions for at least the given percentage of the available logical space",
+	"Consider suggestions for at least the given percentage of the available logical space (zero to disable)",
 	0.10, // more than 10% of space will trigger
 	validateFraction,
 )

